crates/rpc/src/v03/method/simulate_transaction.rs

```diff
@@ -0,0 +1,665 @@
+use crate::{
+    cairo::ext_py::{
+        types::{FeeEstimate, FunctionInvocation, TransactionSimulation, TransactionTrace},
+        CallFailure, GasPriceSource,
+    },
+    context::RpcContext,
+    error::RpcError,
+    v02::{
+        method::estimate_fee::base_block_and_pending_for_call,
+        types::{reply, request::BroadcastedTransaction},
+    },
+};
+
+use anyhow::anyhow;
+use pathfinder_common::BlockId;
+use serde::{Deserialize, Serialize};
+use stark_hash::Felt;
+
+pub async fn simulate_transaction(
+    context: RpcContext,
+    input: SimulateTrasactionInput,
+) -> Result<SimulateTransactionResult, SimulateTransactionError> {
+    let handle = context.call_handle.as_ref().ok_or_else(|| {
+        SimulateTransactionError::Internal(anyhow!("Illegal state: missing call_handle"))
+    })?;
+
+    let gas_price = if matches!(input.block_id, BlockId::Pending | BlockId::Latest) {
+        let gas_price = match context.eth_gas_price.as_ref() {
+            Some(cached) => cached.get().await,
+            None => None,
+        };
+
+        let gas_price =
+            gas_price.ok_or_else(|| anyhow::anyhow!("Current eth_gasPrice is unavailable"))?;
+
+        GasPriceSource::Current(gas_price)
+    } else {
+        GasPriceSource::PastBlock
+    };
+
+    let (at_block, pending_timestamp, pending_update) =
+        base_block_and_pending_for_call(input.block_id, &context.pending_data)
+            .await
+            .map_err(|_| SimulateTransactionError::BlockNotFound)?;
+
+    let skip_execute = input
+        .simulation_flags
+        .0
+        .iter()
+        .any(|flag| flag == &dto::SimulationFlag::SkipExecute);
+    let skip_validate = input
+        .simulation_flags
+        .0
+        .iter()
+        .any(|flag| flag == &dto::SimulationFlag::SkipValidate);
+    let txs = handle
+        .simulate_transaction(
+            at_block,
+            gas_price,
+            pending_update,
+            pending_timestamp,
+            &input.transactions,
+            (skip_execute, skip_validate),
+        )
+        .await
+        .map_err(|e| match e {
+            CallFailure::NoSuchBlock => SimulateTransactionError::BlockNotFound,
+            CallFailure::NoSuchContract => SimulateTransactionError::ContractNotFound,
+            _ => SimulateTransactionError::ContractError,
```

Comment:
I think this leads to us not really returning _any_ information on why execution failed. I think this should be at least handling `ExecutionFailed` explicitly, like in the `call` implementation: https://github.com/eqlabs/pathfinder/blob/8574d4a45b39532aff586210d587caae2cd66cc0/crates/rpc/src/v02/method/call.rs#L13

Reply:
Also, some tests for the error cases would be great.

Reply:
Btw :arrow_up: , can we support [err code 10](https://github.com/starkware-libs/starknet-specs/blob/v0.3.0-rc1/api/starknet_trace_api_openrpc.json#L333)?